File: programmers/graph/가짜-헤밀토니안.py
import sys
# Python의 기본 재귀 깊이 제한을 늘려줍니다 (런타임 에러 해결)
sys.setrecursionlimit(300000)
from collections import defaultdict

def solution(t):
    n = len(t)
    graph = defaultdict(list)
    
    for u, v in t:
        graph[u].append(v)
        graph[v].append(u)
    
    visited = [0] * (n + 1)
    
    # unique_count를 인자로 넘겨 O(1)로 고유 방문자 수를 계산 (시간 초과 해결)
    def check(node, unique_count):
        if visited[node] == 2:
            return 0
        
        # 처음 방문하는 노드일 때만 고유 개수 +1
        if visited[node] == 0:
            unique_count += 1
            
        visited[node] += 1
        
        res = unique_count
        for neigh in graph[node]:
            if visited[neigh] < 2:
                # 다음 노드로 넘어갈 때 갱신된 unique_count를 그대로 전달
                res = max(res, check(neigh, unique_count))
                
        visited[node] -= 1
        return res
    
    res = 0
    # 끝점(Leaf Node)에서만 탐색 시작
    leaf_nodes = [node for node in graph if len(graph[node]) == 1]
    
    # 노드가 1개이거나 간선이 없는 예외 처리
    if not leaf_nodes:
        leaf_nodes = list(graph.keys())
        
    for node in leaf_nodes:
        temp = check(node, 0)
        res = max(res, temp)
        
    return res

File: programmers/graph/가짜-헤밀토니안.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
#include <algorithm>

using namespace std;

unordered_map<long long, int> dfsMemo;
unordered_map<long long, int> funcMemo;

long long makeKey(int cur, int prev) {
    return (static_cast<long long>(cur) << 32) | static_cast<unsigned int>(prev + 1);
}

int dfs(int cur, int prev, unordered_map<int, vector<int>>& graph) {
    long long key = makeKey(cur, prev);
    if (dfsMemo.find(key) != dfsMemo.end()) return dfsMemo[key];

    int depth = 1;
    for (int node : graph[cur]) {
        if (node == prev) continue;
        depth = max(depth, 1 + dfs(node, cur, graph));
    }

    return dfsMemo[key] = depth;
}

int func(int cur, int prev, unordered_map<int, vector<int>>& graph) {
    long long key = makeKey(cur, prev);
    if (funcMemo.find(key) != funcMemo.end()) return funcMemo[key];

    int res = 1;

    int max1 = 0, max2 = 0; // 자식 방향 dfs depth 중 1등, 2등
    for (int node : graph[cur]) {
        if (node == prev) continue;

        int d = dfs(node, cur, graph);
        if (d > max1) {
            max2 = max1;
            max1 = d;
        } else if (d > max2) {
            max2 = d;
        }
    }

    for (int node : graph[cur]) {
        if (node == prev) continue;

        int d = dfs(node, cur, graph);
        int bestOther = (d == max1 ? max2 : max1);

        res = max(res, 1 + bestOther + func(node, cur, graph));
    }

    return funcMemo[key] = res;
}

int solution(vector<vector<int>> t) {
    int n = t.size() + 1;
    unordered_map<int, vector<int>> graph;

    for (auto& arr : t) {
        int u = arr[0], v = arr[1];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }

    dfsMemo.clear();
    funcMemo.clear();

    int res = 0;
    for (int node = 0; node < n; node++) {
        if (graph[node].size() > 1) continue; // leaf만 시작점
        res = max(res, func(node, -1, graph));
    }

    return res;
}
